### PR TITLE
unix: fix race condition in uv_async_send()

### PR DESCRIPTION
**Pull request by bnoordhuis**
After inspection of the code, it seems like there is a race window
between the cmpxchgi() and uv__async_send() calls.

If the event loop thread is already busy looping over the async handles,
it can invoke the callback - which in turn can close the handle - before
the other thread reaches the uv__async_send() call. That's bad because
it accesses the handle that at that point might not be valid anymore.

Fix that by introducing an ad hoc spinlock that blocks the event loop
thread until the sending thread is done. It's not pretty or elegant
but it fixes the immediate bug and appears to have no measurable
impact on the async handle benchmarks.

CI: https://ci.nodejs.org/job/libuv-test-commit/1322/

### CODE REVIEW

**Comment by bnoordhuis** [comment]
> CI: https://ci.nodejs.org/job/libuv-test-commit/1322/

Failures:

1. threadpool_multiple_event_loops on freebsd10-64 but that's a known flake.
2. build failures on the centos6 bots... also happens with other PRs.
```
checking for _FILE_OFFSET_BITS value needed for large files... no
./configure: line 12957: syntax error near unexpected token `;'
./configure: line 12957: `; then'
Build step 'Execute shell' marked build as failure
```
That might have been caused by 5234b1c43a? Here's a CI run with that commit reverted: https://ci.nodejs.org/job/libuv-test-commit/1325/

File: src/unix/async.c [hunk 1]
```diff
@@ -61,14 +61,43 @@ int uv_async_send(uv_async_t* handle) {
   if (ACCESS_ONCE(int, handle->pending) != 0)
```
Comment (santigimeno) [inline src/unix/async.c]: Although unrelated, would somebody care to explain the reason for this `ACCESS_ONCE()` call? Thanks

Comment (bnoordhuis) [reply src/unix/async.c]: cmpxchg is a heavyweight operation compared to a simple load, so this is a fast opt-out.

Comment (santigimeno) [reply src/unix/async.c]: :+1: Thanks for the explanation.

**Comment by josephg** [comment]
Yeah it is a good find. It was first found by @aikoven who saw [crashes running the nodejs bindings of foundationdb every 12 hours or so](https://github.com/apple/foundationdb/issues/1294).

Via [fdb node bindings](https://github.com/josephg/node-foundationdb/issues/28)

Then via [this libuv bug](https://github.com/libuv/libuv/issues/2226)

I'm 99% sure this change will fix the issues we're seeing - though I'd like to confirm with [this repro case](https://gist.github.com/josephg/d646cc4fdbb89ff0a19ec6c5c5550969) to make sure. I might give it a spin tomorrow.

Thanks for putting this together @bnoordhuis - I really appreciate your work and admire your quiet dedication and competence. Not every hero wears a cape.

**Comment by bnoordhuis** [comment]
Thanks for the kind words, @josephg, much appreciated. :-)

**Comment by josephg** [comment]
I can confirm - this PR fixes issue #2226 

😎

**Review by sigusr1 — COMMENTED** [review]
Maybe i misunderstand, anybody can explain follow questions for me, thanks ahead.
If thread A run into uv__async_close, and has called call uv__async_spin, now thread B can call uv_async_send? 
What if handle has been released by thread A?

**Pull request closed without merging**